File: examples/mobile/android/AionMCPModels.kt
package com.example.aionmcp.data.models

import com.google.gson.annotations.SerializedName

/**
 * Health check response from AionMCP server
 */
data class HealthResponse(
    @SerializedName("status")
    val status: String,
    
    @SerializedName("timestamp")
    val timestamp: Long,
    
    @SerializedName("version")
    val version: String,
    
    @SerializedName("iteration")
    val iteration: String? = null
)

/**
 * Tool metadata
 */
data class Tool(
    @SerializedName("name")
    val name: String,
    
    @SerializedName("description")
    val description: String,
    
    @SerializedName("metadata")
    val metadata: Map<String, Any>? = null,
    
    @SerializedName("source")
    val source: String? = null,
    
    @SerializedName("type")
    val type: String? = null
)

/**
 * Response when listing available tools
 */
data class ToolsResponse(
    @SerializedName("protocol")
    val protocol: String,
    
    @SerializedName("tools")
    val tools: List<Tool>
)

/**
 * Request to invoke a tool
 */
data class ToolInvokeRequest(
    @SerializedName("tool")
    val tool: String,
    
    @SerializedName("args")
    val args: Map<String, Any>
)

/**
 * Response from tool invocation
 */
data class ToolInvokeResponse(
    @SerializedName("tool")
    val tool: String,
    
    @SerializedName("result")
    val result: Any
)

/**
 * Learning statistics from the server
 */
data class LearningStatsResponse(
    @SerializedName("total_executions")
    val totalExecutions: Int,
    
    @SerializedName("success_rate")
    val successRate: Double,
    
    @SerializedName("avg_latency_ms")
    val avgLatencyMs: Double,
    
    @SerializedName("tool_stats")
    val toolStats: Map<String, ToolStats>? = null
)

/**
 * Statistics for individual tools
 */
data class ToolStats(
    @SerializedName("executions")
    val executions: Int,
    
    @SerializedName("success_rate")
    val successRate: Double,
    
    @SerializedName("avg_latency_ms")
    val avgLatencyMs: Double
)

/**
 * Learning insight
 */
data class Insight(
    @SerializedName("id")
    val id: String,
    
    @SerializedName("type")
    val type: String,
    
    @SerializedName("priority")
    val priority: String,
    
    @SerializedName("title")
    val title: String,
    
    @SerializedName("description")
    val description: String,
    
    @SerializedName("recommendation")
    val recommendation: String? = null,
    
    @SerializedName("created_at")
    val createdAt: String
)

/**
 * Insights response
 */
data class InsightsResponse(
    @SerializedName("insights")
    val insights: List<Insight>
)

/**
 * API error response
 */
data class ErrorResponse(
    @SerializedName("error")
    val error: String,
    
    @SerializedName("code")
    val code: String? = null,
    
    @SerializedName("details")
    val details: Map<String, Any>? = null
)

File: examples/mobile/ios/AionMCPModels.swift
import Foundation

// MARK: - Health Response

struct HealthResponse: Codable {
    let status: String
    let timestamp: Int64
    let version: String
    let iteration: String?
}

// MARK: - Tool

struct Tool: Codable, Identifiable {
    var id: String { name }
    
    let name: String
    let description: String
    let metadata: [String: AnyCodable]?
    let source: String?
    let type: String?
}

// MARK: - Tools Response

struct ToolsResponse: Codable {
    let `protocol`: String
    let tools: [Tool]
}

// MARK: - Tool Invoke Request

struct ToolInvokeRequest: Codable {
    let tool: String
    let args: [String: AnyCodable]
}

// MARK: - Tool Invoke Response

struct ToolInvokeResponse: Codable {
    let tool: String
    let result: AnyCodable
}

// MARK: - Learning Stats Response

struct LearningStatsResponse: Codable {
    let totalExecutions: Int
    let successRate: Double
    let avgLatencyMs: Double
    let toolStats: [String: ToolStats]?
    
    enum CodingKeys: String, CodingKey {
        case totalExecutions = "total_executions"
        case successRate = "success_rate"
        case avgLatencyMs = "avg_latency_ms"
        case toolStats = "tool_stats"
    }
}

// MARK: - Tool Stats

struct ToolStats: Codable {
    let executions: Int
    let successRate: Double
    let avgLatencyMs: Double
    
    enum CodingKeys: String, CodingKey {
        case executions
        case successRate = "success_rate"
        case avgLatencyMs = "avg_latency_ms"
    }
}

// MARK: - Insight

struct Insight: Codable, Identifiable {
    let id: String
    let type: String
    let priority: String
    let title: String
    let description: String
    let recommendation: String?
    let createdAt: String
    
    enum CodingKeys: String, CodingKey {
        case id, type, priority, title, description, recommendation
        case createdAt = "created_at"
    }
}

// MARK: - Insights Response

struct InsightsResponse: Codable {
    let insights: [Insight]
}

// MARK: - Error Response

struct ErrorResponse: Codable {
    let error: String
    let code: String?
    let details: [String: AnyCodable]?
}

// MARK: - AnyCodable Helper

/// Helper type for encoding/decoding Any values
struct AnyCodable: Codable {
    let value: Any
    
    init(_ value: Any) {
        self.value = value
    }
    
    init(from decoder: Decoder) throws {
        let container = try decoder.singleValueContainer()
        
        if let int = try? container.decode(Int.self) {
            value = int
        } else if let double = try? container.decode(Double.self) {
            value = double
        } else if let string = try? container.decode(String.self) {
            value = string
        } else if let bool = try? container.decode(Bool.self) {
            value = bool
        } else if let array = try? container.decode([AnyCodable].self) {
            value = array.map { $0.value }
        } else if let dict = try? container.decode([String: AnyCodable].self) {
            value = dict.mapValues { $0.value }
        } else {
            value = NSNull()
        }
    }
    
    func encode(to encoder: Encoder) throws {
        var container = encoder.singleValueContainer()
        
        switch value {
        case let int as Int:
            try container.encode(int)
        case let int64 as Int64:
            try container.encode(int64)
        case let double as Double:
            try container.encode(double)
        case let float as Float:
            try container.encode(float)
        case let string as String:
            try container.encode(string)
        case let bool as Bool:
            try container.encode(bool)
        case let array as [Any]:
            try container.encode(array.map { AnyCodable($0) })
        case let dict as [String: Any]:
            try container.encode(dict.mapValues { AnyCodable($0) })
        case is NSNull:
            try container.encodeNil()
        default:
            try container.encodeNil()
        }
    }
}

// MARK: - Custom Errors

enum AionMCPError: LocalizedError {
    case networkError(Error)
    case apiError(String)
    case decodingError(Error)
    case invalidResponse
    case unauthorized
    case serverError(Int)
    
    var errorDescription: String? {
        switch self {
        case .networkError(let error):
            return "Network error: \(error.localizedDescription)"
        case .apiError(let message):
            return "API error: \(message)"
        case .decodingError(let error):
            return "Failed to decode response: \(error.localizedDescription)"
        case .invalidResponse:
            return "Invalid response from server"
        case .unauthorized:
            return "Unauthorized access"
        case .serverError(let code):
            return "Server error with code: \(code)"
        }
    }
}
